src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,287 @@
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "zcash/IncrementalMerkleTree.hpp"
+#include "crypto/sha256.h"
+#include "zerocash/utils/util.h" // TODO: remove these utilities
+
+namespace libzcash {
+
+SHA256Compress SHA256Compress::combine(const SHA256Compress& a, const SHA256Compress& b)
+{
+    // This is a performance optimization.
+    if (a.IsNull() && b.IsNull()) {
+        return a;
+    }
+
+    SHA256Compress res = SHA256Compress();
+
+    CSHA256 hasher;
+    hasher.Write(a.begin(), 32);
+    hasher.Write(b.begin(), 32);
+    hasher.FinalizeNoPadding(res.begin());
+
+    return res;
+}
+
+
+template <typename Hash>
+class PathFiller {
+private:
+    std::deque<Hash> queue;
+public:
+    PathFiller() : queue() { }
+    PathFiller(std::deque<Hash> queue) : queue(queue) { }
+
+    Hash next() {
+        if (queue.size() > 0) {
+            Hash h = queue.front();
+            queue.pop_front();
+
+            return h;
+        } else {
+            return Hash();
+        }
+    }
+};
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::wfcheck() const {
+    if (parents.size() >= Depth) {
+        throw std::ios_base::failure("tree has too many parents");
+    }
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::append(Hash obj) {
+    if (is_complete(Depth)) {
+        throw std::runtime_error("tree is full");
+    }
+
+    if (!left) {
+        // Set the left leaf
+        left = obj;
+    } else if (!right) {
+        // Set the right leaf
+        right = obj;
+    } else {
+        // Combine the leaves and propagate it up the tree
+        boost::optional<Hash> combined = Hash::combine(*left, *right);
+
+        // Set the left leaf to the object and make the right object none
+        left = obj;
+        right = boost::none;
+
+        // Propagate up the tree as much as needed
+        BOOST_FOREACH(boost::optional<Hash>& parent, parents) {
+            if (parent) {
+                combined = Hash::combine(*parent, *combined);
+                parent = boost::none;
+            } else {
+                parent = *combined;
+                combined = boost::none;
+                break;
+            }
+        }
+
+        if (combined) {
+            // Create a new parent
+            parents.push_back(combined);
+        }
+    }
+}
```

Comment:
One way to rewrite this with less cases would be:

Initialization:

```
Roots = [].
```

Append(obj):

```
for (i = 0; i < depth; i++):
    if i < roots.length() && roots[i]:
        obj = H(roots[i] || obj)
        roots[i] = none
    else
        /* TODO: grow roots if necessary */
        roots[i] = obj;
    endif
endfor
```

That's equivalent because `roots[0]` is the same as `left` in the current code and is thought about as "the left depth-0 subtree of the rightmost depth-1 subtree". If we did this then we wouldn't be keeping `right` around (as soon as you add a right child it gets hashed). Maybe keeping `right` around is necessary for some of the code I haven't read yet.